teuthology/scrape.py

```diff
@@ -0,0 +1,519 @@
+
+
+class Reason(object):
+    def get_description(self):
+        return self.description
+
+    def get_detail(self):
+        return None
+
+
+def grep(path, expr):
+    """
+    Call out to native grep rather than feeding massive log files through python line by line
+    """
+    p = subprocess.Popen(["grep", expr, path], stdout=subprocess.PIPE)
+    out, err = p.communicate()
+    if p.returncode == 0:
+        return out.split("\n")
+    else:
+        return []
+
+
+class GenericReason(Reason):
+    """
+    A reason inferred from a Job: matches Jobs with an apparently-similar failure
+    """
+
+    def __init__(self, job, description=None):
+        self.failure_reason = job.get_failure_reason()
+        self.description = description
+
+        self.backtrace = job.get_backtrace()
+        if self.backtrace:
+            log.debug("Found a backtrace!\n{0}".format(self.backtrace))
+
+    def get_detail(self):
+        return self.backtrace
+
+    def get_description(self):
+        if self.description:
+            return self.description
+        else:
+            if self.backtrace:
+                return "Crash: {0}".format(self.failure_reason)
+            else:
+                return "Failure: {0}".format(self.failure_reason)
+
+    def match(self, job):
+        # I never match dead jobs
+        if job.get_failure_reason() is None:
+            return False
+
+        # If one has a backtrace but the other doesn't, we're a different thing even if the official
+        # failure_reason is the same
+        if (self.backtrace is None) != (job.get_backtrace() is None):
+            return False
+
+        # If we have the same backtrace, we're a match even if the teuthology failure_reason
+        # doesn't match (a crash is a crash, it can have different symptoms)
+        if self.backtrace:
+            ratio = difflib.SequenceMatcher(None, self.backtrace, job.get_backtrace()).ratio()
+            return ratio > 0.5
+        else:
+            if "Test failure:" in self.failure_reason:
+                return self.failure_reason == job.get_failure_reason()
+            elif re.search("workunit test (.*)\) on ", self.failure_reason):
+                workunit_name = re.search("workunit test (.*)\) on ", self.failure_reason).group(1)
+                other_match = re.search("workunit test (.*)\) on ", job.get_failure_reason())
+                return other_match is not None and workunit_name == other_match.group(1)
+            else:
+                reason_ratio = difflib.SequenceMatcher(None, self.failure_reason, job.get_failure_reason()).ratio()
+                return reason_ratio > 0.5
+
+
+class RegexReason(Reason):
+    """
+    A known reason matching a particular regex to failure reason
+    """
+
+    def __init__(self, regexes, description):
+        self.description = description
+        if not isinstance(regexes, list):
+            self.regexes = [regexes]
+        else:
+            self.regexes = regexes
```

Comment:
Can this have change check order to avoid negative condition in if for straight logic.

Reply:
Do you mean just flipping the lines of if and else?

Reply:
Right, like:
```
if isinstance(regexes, list):
    self.regexes = regexes
else:
    self.regexes = [regexes]
```